pkg/pipeline/utils/exit.go

```diff
@@ -0,0 +1,42 @@
+package utils
+
+import (
+	log "github.com/sirupsen/logrus"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+)
+
+var (
+	registeredChannels []chan bool
+	chanMutex          sync.Mutex
+)
+
+func RegisterExitChannel(ch chan bool) {
+	chanMutex.Lock()
+	registeredChannels = append(registeredChannels, ch)
+	chanMutex.Unlock()
+}
+
+func SetupElegantExit() {
+	log.Debugf("entering SetupElegantExit")
+	// handle elegant exit; create support for channels of go routines that want to exit cleanly
+	registeredChannels = make([]chan bool, 0)
+	exitSigChan := make(chan os.Signal, 1)
+	log.Debugf("registered exit signal channel")
+	signal.Notify(exitSigChan, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		// wait for exit signal; then stop all the other go functions
+		sig := <-exitSigChan
+		log.Debugf("received exit signal = %v", sig)
+		chanMutex.Lock()
+		// exit signal received; stop other go functions
+		for _, ch := range registeredChannels {
+			ch <- true
+		}
+		chanMutex.Unlock()
+		log.Debugf("exiting SetupElegantExit go function")
```

Comment:
How do we know the order of things here ... I understand that you notify all registered channels ... and that they will exit ... but when will the entire process exit??? maybe you need to wait a sec or two to make sure all registeredChannels have a chance to exit

Reply:
added a second in main before exit.